### PR TITLE
Replace pointwise convolutions with matmul in ConformerEncoder: 10% E2E training speedup

### PR DESCRIPTION
**Pull request by MahmoudAshraf97**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

## Problem

`ConformerConvolution`'s two `nn.Conv1d(kernel_size=1)` projections lower to `cutlass_80_wmma_tensorop_bf16_s161616gemm_32x32` (old WMMA API) on an RTX 5090 / bf16, instead of the `cutlass_80_tensorop_bf16_s16816gemm_64x128` mma.sync kernel the same math gets as a linear which is 2.6-2.8x slower on those ops. `torch.compile` doesn't improve it.

## Fix

Apply the same weights with `nn.functional.linear`. Modules stay `nn.Conv1d`, so `state_dict` and checkpoints are untouched. Numerically identical (8.4e-07); streaming caches and export unaffected.

## Benchmarks

Encoder fwd+bwd, RTX 5090, bf16, **audio-seconds per wall-second**. Batch identical in both arms.

### Eager

| model | audio | batch | main | this PR | speedup |
|---|---|---|---|---|---|
| M (d=256, L=16) | 10.2 s | 207 | 10,115 | 10,814 | 1.069x |
| | 20.5 s | 75 | 8,577 | 8,984 | 1.047x |
| | 34.6 s | 36 | 7,327 | 7,618 | 1.040x |
| | 60.2 s | 17 | 5,581 | 5,745 | 1.029x |
| L (d=512, L=17) | 10.2 s | 103 | 5,629 | 6,285 | 1.117x |
| | 20.5 s | 37 | 4,751 | 5,150 | 1.084x |
| | 34.6 s | 18 | 3,967 | 4,242 | 1.069x |
| | 60.2 s | 8 | 2,950 | 3,091 | 1.048x |
| XL (d=1024, L=24) | 10.2 s | 51 | 2,000 | 2,255 | 1.127x |
| | 20.5 s | 18 | 1,913 | 2,118 | 1.107x |
| | 34.6 s | 9 | 1,698 | 1,909 | 1.124x |
| | 60.2 s | 4 | 1,357 | 1,483 | 1.093x |
| **aggregate** | | | **4,625** | **4,999** | **1.081x** |

### torch.compile

| model | audio | batch | main | this PR | speedup |
|---|---|---|---|---|---|
| M (d=256, L=16) | 10.2 s | 207 | 15,551 | 16,500 | 1.061x |
| | 20.5 s | 75 | 15,464 | 16,282 | 1.053x |
| | 34.6 s | 36 | 14,254 | 15,169 | 1.064x |
| | 60.2 s | 17 | 12,151 | 13,126 | 1.080x |
| L (d=512, L=17) | 10.2 s | 103 | 7,908 | 8,848 | 1.119x |
| | 20.5 s | 37 | 7,527 | 8,267 | 1.098x |
| | 34.6 s | 18 | 7,122 | 7,940 | 1.115x |
| | 60.2 s | 8 | 5,846 | 6,467 | 1.106x |
| XL (d=1024, L=24) | 10.2 s | 51 | 2,399 | 2,683 | 1.119x |
| | 20.5 s | 18 | 2,376 | 2,632 | 1.108x |
| | 34.6 s | 9 | 2,212 | 2,514 | 1.137x |
| | 60.2 s | 4 | 1,901 | 2,132 | 1.122x |
| **aggregate** | | | **7,086** | **7,803** | **1.101x** |

## Testing

ASR encoder/export tests pass; TorchScript export verified with `check_trace=True`. Note `pointwise_conv{1,2}.forward` are no longer called, so forward hooks on those two convs won't fire.

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?
@pzelasko 

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Review by pzelasko — DISMISSED** [review]
Thank you, that's a cool contribution! I don't see any blocker from merging as I don't expect us to use hooks extensively on pointwise_conv{1,2} (which would not fire because of explicit F.linear call). I'd like a second opinion from @nithinraok

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 1]
```diff
@@ -336,17 +336,21 @@ def __init__(
         )
 
     def forward(self, x, pad_mask=None, cache=None):
-        x = x.transpose(1, 2)
-        x = self.pointwise_conv1(x)
+        # kernel_size=1 convs are pointwise, i.e. linear, but nn.Conv1d dispatches to much slower
+        # cuBLAS kernels; the modules stay nn.Conv1d so checkpoints are unchanged.
+        x = nn.functional.linear(x, self.pointwise_conv1.weight.squeeze(-1), self.pointwise_conv1.bias)
```
Comment (nithinraok) [inline nemo/collections/asr/parts/submodules/conformer_modules.py]: Could you move operation to something like this to make operation simpler looking:
```
  def _linear_from_pointwise_conv(self, x, conv):
      return nn.functional.linear(x, conv.weight.squeeze(-1), conv.bias)
```

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 2]
```diff
@@ -336,17 +336,21 @@ def __init__(
...
-            x = nn.functional.glu(x, dim=1)
+            x = nn.functional.glu(x, dim=-1)
         else:
             x = self.pointwise_activation(x)
 
+        # the depthwise conv and its streaming cache are channel-first
+        x = x.transpose(1, 2)
         if pad_mask is not None:
             x = x.masked_fill(pad_mask.unsqueeze(1), 0.0)
+        x = x.contiguous()
```
Comment (nithinraok) [inline nemo/collections/asr/parts/submodules/conformer_modules.py]: do we need this copy?

**Review by nithinraok — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/parts/submodules/conformer_modules.py (hunk 2 above)
Comment (MahmoudAshraf97) [reply nemo/collections/asr/parts/submodules/conformer_modules.py]: No, but I assumed that the linear will have worse performance with strided input, but it handled it well, so I removed it

File: nemo/collections/asr/parts/submodules/conformer_modules.py (hunk 1 above)
Comment (MahmoudAshraf97) [reply nemo/collections/asr/parts/submodules/conformer_modules.py]: addressed and moved the explanatory comment there for clarity, but renamed it to `_apply_pointwise` as the suggested name implied that the return is a layer instead of a tensor

**Review by nithinraok — APPROVED** [review]
Thanks!
